tmt/steps/prepare/feature/__init__.py

```diff
@@ -32,51 +70,37 @@ def __init__(
 
         self.guest = guest
 
-    def _find_playbook(self, filename: str) -> Optional[Path]:
+    @classmethod
+    def enable(cls, guest: Guest, logger: tmt.log.Logger) -> None:
+        raise NotImplementedError
+
+    @classmethod
+    def disable(cls, guest: Guest, logger: tmt.log.Logger) -> None:
+        raise NotImplementedError
+
+    @classmethod
+    def _find_playbook(cls, filename: str, logger: tmt.log.Logger) -> Optional[Path]:
         filepath = FEATURE_PLAYEBOOK_DIRECTORY / filename
         if filepath.exists():
             return filepath
 
-        self.warn(f"Cannot find any suitable playbook for '{filename}'.")
+        logger.warning(f"Cannot find any suitable playbook for '{filename}'.", 0)
         return None
 
-
-class ToggleableFeature(Feature):
-    def _run_playbook(self, op: str, playbook_filename: str) -> None:
-        playbook_path = self._find_playbook(playbook_filename)
+    @classmethod
+    def _run_playbook(
+            cls,
+            op: str,
+            playbook_filename: str,
+            guest: Guest,
+            logger: tmt.log.Logger) -> None:
+        playbook_path = cls._find_playbook(playbook_filename, logger)
         if not playbook_path:
             raise tmt.utils.GeneralError(
-                f"{op.capitalize()} {self.NAME.upper()} is not supported on this guest.")
-
-        self.info(f'{op.capitalize()} {self.NAME.upper()}')
-        self.guest.ansible(playbook_path)
-
-    def _enable(self, playbook_filename: str) -> None:
-        self._run_playbook('enable', playbook_filename)
-
-    def _disable(self, playbook_filename: str) -> None:
-        self._run_playbook('disable', playbook_filename)
-
-    def enable(self) -> None:
-        raise NotImplementedError
-
-    def disable(self) -> None:
-        raise NotImplementedError
-
-
-class EPEL(ToggleableFeature):
-    NAME = 'epel'
+                f"{op.capitalize()} {cls.NAME.upper()} is not supported on this guest.")
 
-    def enable(self) -> None:
-        self._enable('epel-enable.yaml')
-
-    def disable(self) -> None:
-        self._disable('epel-disable.yaml')
-
-
-_FEATURES: dict[str, type[Feature]] = {
-    EPEL.NAME: EPEL
-    }
+        logger.info(f'{op.capitalize()} {cls.NAME.upper()}')
+        guest.ansible(playbook_path)
 
 
 @dataclasses.dataclass
```

Comment:
Dont you mean this? https://github.com/teemtee/tmt/blob/cd8f11e2653430640384c2b6ed9c0fb49ba9840c/tmt/steps/prepare/feature/__init__.py#L107
Here we can add other custom fields?

Reply:
I think I now get it, that we currently use PrepareStepData but we would like to have PluginStepData?

Reply:
We can, but the question is why would we add the custom field? What drives us to add a new field? It's a new feature plugin. But there can be out-of-the-tree plugins, plugins developed by users and teams. How would these plugins add the fields they need into `PrepareFeatureData`? It cannot be added by us, manually, by updating `feature.py`, because then what would be the point of developing a custom plugin for your team when you need to talk to tmt folks and get them "allocate" or "reserve" a key in this container for your plugin.

The answer is, that this must be based on discovery: tmt discovers `feature` plugins, and asks them, nicely, "Are there any fields you wish to add to `PrepareFeatureData`?". And plugins would then return fields they wish to make available: `epel` would add the `epel` field that's now hardcoded in this container, and so on. Basically `PrepareFeatureData` content would be generated dynamically as tmt starts and discovers plugins. And for exactly this missing piece I'd like to see a TODO, because we can resolve it later :)

Reply:
Yeah, definitely makes sense. And the `TODO` is there. So, I guess, this part is done for now?